File: src/attractor/engine/context_contracts.py
from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
import json
import re
from typing import Collection, Mapping

from attractor.dsl.models import DotAttribute


_ALLOWED_CONTEXT_UPDATE_PREFIXES: tuple[str, ...] = (
    "context.",
    "graph.",
    "internal.",
    "parallel.",
    "stack.",
    "human.gate.",
    "work.",
    "_attractor.",
)
_CONTEXT_READ_SHORTHAND_PREFIXES: tuple[str, ...] = (
    "request.",
    "review.",
    "planflow.",
    "milestone.",
    "item.",
)
_CONTEXT_UPDATE_KEY_RE = re.compile(r"^[A-Za-z0-9_-]+(?:\.[A-Za-z0-9_-]+)*$")


@dataclass(frozen=True)
class ContextWriteContract:
    allowed_keys: tuple[str, ...] = ()
    parse_error: str = ""


@dataclass(frozen=True)
class ContextReadContract:
    declared_keys: tuple[str, ...] = ()
    parse_error: str = ""


@dataclass(frozen=True)
class ContextWriteContractViolation:
    offending_keys: tuple[str, ...]
    allowed_keys: tuple[str, ...]
    parse_error: str = ""
    invalid_keys: tuple[str, ...] = ()

    def format_reason(self, *, node_id: str | None = None) -> str:
        node_prefix = f" for node '{node_id}'" if node_id else ""
        offending = ", ".join(self.offending_keys) if self.offending_keys else "<none>"
        declared = ", ".join(self.allowed_keys) if self.allowed_keys else "<none>"
        invalid = ", ".join(self.invalid_keys) if self.invalid_keys else "<none>"
        parts = []
        if self.invalid_keys:
            parts.append(f"invalid context_updates keys{node_prefix}: {invalid}")
            parts.append("context_updates keys must be bare identifiers or dot-separated identifiers")
        parts.extend(
            [
                f"undeclared context_updates keys{node_prefix}: {offending}",
                f"declared spark.writes_context allowlist: {declared}",
            ]
        )
        if self.parse_error:
            parts.append(f"spark.writes_context parse error: {self.parse_error}")
        return "; ".join(parts)


def normalize_context_key(key: str) -> str:
    normalized = str(key).strip()
    if not normalized or "." not in normalized:
        return normalized
    if _invalid_context_key_reason(normalized):
        return normalized
    if any(normalized.startswith(prefix) for prefix in _ALLOWED_CONTEXT_UPDATE_PREFIXES):
        return normalized
    return f"context.{normalized}"


def normalize_context_update_key(key: str) -> str:
    return normalize_context_key(key)


def normalize_context_read_key(key: str) -> str:
    normalized = str(key).strip()
    if not normalized or "." not in normalized:
        return normalized
    if _invalid_context_key_reason(normalized):
        return normalized
    if any(normalized.startswith(prefix) for prefix in _ALLOWED_CONTEXT_UPDATE_PREFIXES):
        return normalized
    if any(normalized.startswith(prefix) for prefix in _CONTEXT_READ_SHORTHAND_PREFIXES):
        return f"context.{normalized}"
    return normalized


def resolve_context_write_contract(node_attrs: Mapping[str, DotAttribute]) -> ContextWriteContract:
    attr = node_attrs.get("spark.writes_context")
    raw_value = None if attr is None else str(attr.value)
    return _parse_context_write_contract(raw_value)


def resolve_context_read_contract(node_attrs: Mapping[str, DotAttribute]) -> ContextReadContract:
    attr = node_attrs.get("spark.reads_context")
    raw_value = None if attr is None else str(attr.value)
    return _parse_context_read_contract(raw_value)


def validate_context_updates_against_contract(
    context_updates: Mapping[str, object],
    contract: ContextWriteContract,
    *,
    exempt_keys: Collection[str] = (),
    exempt_prefixes: Collection[str] = (),
) -> ContextWriteContractViolation | None:
    normalized_authored_keys: set[str] = set()
    invalid_authored_keys: set[str] = set()
    for raw_key in context_updates.keys():
        raw_key_text = str(raw_key).strip()
        normalized_key = normalize_context_update_key(raw_key_text)
        if _is_exempt_context_update_key(
            normalized_key,
            exempt_keys=exempt_keys,
            exempt_prefixes=exempt_prefixes,
        ):
            continue
        invalid_reason = _invalid_context_key_reason(raw_key_text)
        if invalid_reason:
            invalid_authored_keys.add(raw_key_text)
            continue
        normalized_authored_keys.add(normalized_key)
    if not normalized_authored_keys and not invalid_authored_keys:
        return None

    allowed = set(contract.allowed_keys)
    offending = tuple(sorted(key for key in normalized_authored_keys if key not in allowed))
    invalid_keys = tuple(sorted(invalid_authored_keys))
    if not offending and not invalid_keys and not contract.parse_error:
        return None
    if contract.parse_error and not offending and not invalid_keys:
        offending = tuple(sorted(normalized_authored_keys))
    return ContextWriteContractViolation(
        offending_keys=offending,
        allowed_keys=contract.allowed_keys,
        parse_error=contract.parse_error,
        invalid_keys=invalid_keys,
    )


def _is_exempt_context_update_key(
    key: str,
    *,
    exempt_keys: Collection[str],
    exempt_prefixes: Collection[str],
) -> bool:
    if key in exempt_keys:
        return True
    return any(key.startswith(prefix) for prefix in exempt_prefixes)


@lru_cache(maxsize=None)
def _parse_context_write_contract(raw_value: str | None) -> ContextWriteContract:
    parsed_contract = _parse_context_key_contract(
        raw_value,
        normalize_key=normalize_context_update_key,
        empty_key_error="expected non-empty context update keys",
        invalid_key_label="context update key",
    )
    return ContextWriteContract(
        allowed_keys=parsed_contract.keys,
        parse_error=parsed_contract.parse_error,
    )


@lru_cache(maxsize=None)
def _parse_context_read_contract(raw_value: str | None) -> ContextReadContract:
    parsed_contract = _parse_context_key_contract(
        raw_value,
        normalize_key=normalize_context_read_key,
        empty_key_error="expected non-empty context read keys",
        invalid_key_label="context read key",
    )
    return ContextReadContract(
        declared_keys=parsed_contract.keys,
        parse_error=parsed_contract.parse_error,
    )


@dataclass(frozen=True)
class _ParsedContextKeyContract:
    keys: tuple[str, ...] = ()
    parse_error: str = ""


def _parse_context_key_contract(
    raw_value: str | None,
    *,
    normalize_key,
    empty_key_error: str,
    invalid_key_label: str,
) -> _ParsedContextKeyContract:
    normalized_raw = "" if raw_value is None else raw_value.strip()
    if not normalized_raw:
        return _ParsedContextKeyContract()

    try:
        parsed = json.loads(normalized_raw)
    except json.JSONDecodeError as exc:
        return _ParsedContextKeyContract(
            parse_error=f"expected a JSON array of strings: {exc.msg} at line {exc.lineno} column {exc.colno}"
        )

    if not isinstance(parsed, list):
        return _ParsedContextKeyContract(parse_error="expected a JSON array of strings")
    if any(not isinstance(item, str) for item in parsed):
        return _ParsedContextKeyContract(parse_error="expected a JSON array of strings")

    normalized_keys: set[str] = set()
    for item in parsed:
        normalized_item = str(item).strip()
        if not normalized_item:
            return _ParsedContextKeyContract(parse_error=empty_key_error)
        invalid_reason = _invalid_context_key_reason(normalized_item)
        if invalid_reason:
            return _ParsedContextKeyContract(
                parse_error=f"invalid {invalid_key_label} '{normalized_item}': {invalid_reason}"
            )
        normalized_item = normalize_key(normalized_item)
        if not normalized_item:
            return _ParsedContextKeyContract(parse_error=empty_key_error)
        normalized_keys.add(normalized_item)
    return _ParsedContextKeyContract(keys=tuple(sorted(normalized_keys)))


def _invalid_context_key_reason(key: str) -> str:
    normalized = str(key).strip()
    if not normalized:
        return "keys must be non-empty"
    if "/" in normalized or "\\" in normalized:
        return "path separators are not allowed"
    if normalized.startswith(".") or normalized.endswith(".") or ".." in normalized:
        return "empty dotted segments are not allowed"
    if not _CONTEXT_UPDATE_KEY_RE.fullmatch(normalized):
        return "keys must be bare identifiers or dot-separated identifiers"
    return ""

File: crates/attractor-core/src/context_contracts.rs
use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};

use crate::context::ContextMap;
use crate::graph::{attr_string, DotAttribute};

const ALLOWED_CONTEXT_UPDATE_PREFIXES: &[&str] = &[
    "context.",
    "graph.",
    "internal.",
    "parallel.",
    "stack.",
    "human.gate.",
    "work.",
    "_attractor.",
];

const CONTEXT_READ_SHORTHAND_PREFIXES: &[&str] =
    &["request.", "review.", "planflow.", "milestone.", "item."];

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ContextWriteContract {
    #[serde(default)]
    pub allowed_keys: Vec<String>,
    #[serde(default)]
    pub parse_error: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ContextReadContract {
    #[serde(default)]
    pub declared_keys: Vec<String>,
    #[serde(default)]
    pub parse_error: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ContextWriteContractViolation {
    pub offending_keys: Vec<String>,
    pub allowed_keys: Vec<String>,
    #[serde(default)]
    pub parse_error: String,
    #[serde(default)]
    pub invalid_keys: Vec<String>,
}

impl ContextWriteContractViolation {
    pub fn format_reason(&self, node_id: Option<&str>) -> String {
        let node_prefix = node_id
            .map(|id| format!(" for node '{id}'"))
            .unwrap_or_default();
        let offending = join_or_none(&self.offending_keys);
        let declared = join_or_none(&self.allowed_keys);
        let invalid = join_or_none(&self.invalid_keys);
        let mut parts = Vec::new();
        if !self.invalid_keys.is_empty() {
            parts.push(format!(
                "invalid context_updates keys{node_prefix}: {invalid}"
            ));
            parts.push(
                "context_updates keys must be bare identifiers or dot-separated identifiers"
                    .to_string(),
            );
        }
        parts.push(format!(
            "undeclared context_updates keys{node_prefix}: {offending}"
        ));
        parts.push(format!(
            "declared spark.writes_context allowlist: {declared}"
        ));
        if !self.parse_error.is_empty() {
            parts.push(format!(
                "spark.writes_context parse error: {}",
                self.parse_error
            ));
        }
        parts.join("; ")
    }
}

pub fn normalize_context_update_key(key: &str) -> String {
    normalize_context_key(key)
}

pub fn normalize_context_read_key(key: &str) -> String {
    let normalized = key.trim();
    if normalized.is_empty() || !normalized.contains('.') {
        return normalized.to_string();
    }
    if invalid_context_key_reason(normalized).is_some() {
        return normalized.to_string();
    }
    if starts_with_any(normalized, ALLOWED_CONTEXT_UPDATE_PREFIXES) {
        return normalized.to_string();
    }
    if starts_with_any(normalized, CONTEXT_READ_SHORTHAND_PREFIXES) {
        return format!("context.{normalized}");
    }
    normalized.to_string()
}

pub fn parse_context_write_contract(raw: Option<&str>) -> ContextWriteContract {
    let parsed = parse_context_key_contract(
        raw,
        normalize_context_update_key,
        "expected non-empty context update keys",
        "context update key",
    );
    ContextWriteContract {
        allowed_keys: parsed.keys,
        parse_error: parsed.parse_error,
    }
}

pub fn parse_context_read_contract(raw: Option<&str>) -> ContextReadContract {
    let parsed = parse_context_key_contract(
        raw,
        normalize_context_read_key,
        "expected non-empty context read keys",
        "context read key",
    );
    ContextReadContract {
        declared_keys: parsed.keys,
        parse_error: parsed.parse_error,
    }
}

pub fn resolve_context_write_contract(
    attrs: &BTreeMap<String, DotAttribute>,
) -> ContextWriteContract {
    let raw_value = attrs
        .get("spark.writes_context")
        .map(|_| attr_string(attrs, "spark.writes_context"));
    parse_context_write_contract(raw_value.as_deref())
}

pub fn resolve_context_read_contract(
    attrs: &BTreeMap<String, DotAttribute>,
) -> ContextReadContract {
    let raw_value = attrs
        .get("spark.reads_context")
        .map(|_| attr_string(attrs, "spark.reads_context"));
    parse_context_read_contract(raw_value.as_deref())
}

pub fn normalize_context_updates(updates: &ContextMap) -> ContextMap {
    updates
        .iter()
        .map(|(key, value)| (normalize_context_update_key(key), value.clone()))
        .collect()
}

pub fn validate_context_updates_against_contract(
    updates: &ContextMap,
    contract: &ContextWriteContract,
) -> Option<ContextWriteContractViolation> {
    validate_context_updates_against_contract_with_exemptions(updates, contract, &[], &[])
}

pub fn validate_context_updates_against_contract_with_exemptions(
    updates: &ContextMap,
    contract: &ContextWriteContract,
    exempt_keys: &[&str],
    exempt_prefixes: &[&str],
) -> Option<ContextWriteContractViolation> {
    let mut normalized_authored_keys = BTreeSet::new();
    let mut invalid_authored_keys = BTreeSet::new();

    for raw_key in updates.keys() {
        let raw_key_text = raw_key.trim();
        let normalized_key = normalize_context_update_key(raw_key_text);
        if is_exempt_context_update_key(&normalized_key, exempt_keys, exempt_prefixes) {
            continue;
        }
        if invalid_context_key_reason(raw_key_text).is_some() {
            invalid_authored_keys.insert(raw_key_text.to_string());
            continue;
        }
        normalized_authored_keys.insert(normalized_key);
    }

    if normalized_authored_keys.is_empty() && invalid_authored_keys.is_empty() {
        return None;
    }

    let allowed: BTreeSet<_> = contract.allowed_keys.iter().cloned().collect();
    let mut offending_keys: Vec<String> = normalized_authored_keys
        .iter()
        .filter(|key| !allowed.contains(*key))
        .cloned()
        .collect();
    let invalid_keys: Vec<String> = invalid_authored_keys.into_iter().collect();

    if offending_keys.is_empty() && invalid_keys.is_empty() && contract.parse_error.is_empty() {
        return None;
    }
    if !contract.parse_error.is_empty() && offending_keys.is_empty() && invalid_keys.is_empty() {
        offending_keys = normalized_authored_keys.into_iter().collect();
    }

    Some(ContextWriteContractViolation {
        offending_keys,
        allowed_keys: contract.allowed_keys.clone(),
        parse_error: contract.parse_error.clone(),
        invalid_keys,
    })
}

fn normalize_context_key(key: &str) -> String {
    let normalized = key.trim();
    if normalized.is_empty() || !normalized.contains('.') {
        return normalized.to_string();
    }
    if invalid_context_key_reason(normalized).is_some() {
        return normalized.to_string();
    }
    if starts_with_any(normalized, ALLOWED_CONTEXT_UPDATE_PREFIXES) {
        return normalized.to_string();
    }
    format!("context.{normalized}")
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
struct ParsedContextKeyContract {
    keys: Vec<String>,
    parse_error: String,
}

fn parse_context_key_contract(
    raw: Option<&str>,
    normalize_key: fn(&str) -> String,
    empty_key_error: &'static str,
    invalid_key_label: &'static str,
) -> ParsedContextKeyContract {
    let normalized_raw = raw.unwrap_or("").trim();
    if normalized_raw.is_empty() {
        return ParsedContextKeyContract::default();
    }

    let parsed: serde_json::Value = match serde_json::from_str(normalized_raw) {
        Ok(value) => value,
        Err(source) => {
            return ParsedContextKeyContract {
                keys: Vec::new(),
                parse_error: format!(
                    "expected a JSON array of strings: {} at line {} column {}",
                    source,
                    source.line(),
                    source.column()
                ),
            };
        }
    };

    let Some(items) = parsed.as_array() else {
        return ParsedContextKeyContract {
            keys: Vec::new(),
            parse_error: "expected a JSON array of strings".to_string(),
        };
    };

    let mut normalized_keys = BTreeSet::new();
    for item in items {
        let Some(text) = item.as_str() else {
            return ParsedContextKeyContract {
                keys: Vec::new(),
                parse_error: "expected a JSON array of strings".to_string(),
            };
        };
        let normalized_item = text.trim();
        if normalized_item.is_empty() {
            return ParsedContextKeyContract {
                keys: Vec::new(),
                parse_error: empty_key_error.to_string(),
            };
        }
        if let Some(reason) = invalid_context_key_reason(normalized_item) {
            return ParsedContextKeyContract {
                keys: Vec::new(),
                parse_error: format!("invalid {invalid_key_label} '{normalized_item}': {reason}"),
            };
        }
        let normalized_item = normalize_key(normalized_item);
        if normalized_item.is_empty() {
            return ParsedContextKeyContract {
                keys: Vec::new(),
                parse_error: empty_key_error.to_string(),
            };
        }
        normalized_keys.insert(normalized_item);
    }

    ParsedContextKeyContract {
        keys: normalized_keys.into_iter().collect(),
        parse_error: String::new(),
    }
}

fn invalid_context_key_reason(key: &str) -> Option<&'static str> {
    let normalized = key.trim();
    if normalized.is_empty() {
        return Some("keys must be non-empty");
    }
    if normalized.contains('/') || normalized.contains('\\') {
        return Some("path separators are not allowed");
    }
    if normalized.starts_with('.') || normalized.ends_with('.') || normalized.contains("..") {
        return Some("empty dotted segments are not allowed");
    }
    if !normalized
        .split('.')
        .all(|segment| segment.chars().all(is_context_key_char))
    {
        return Some("keys must be bare identifiers or dot-separated identifiers");
    }
    None
}

fn is_context_key_char(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_' || ch == '-'
}

fn starts_with_any(value: &str, prefixes: &[&str]) -> bool {
    prefixes.iter().any(|prefix| value.starts_with(prefix))
}

fn is_exempt_context_update_key(key: &str, exempt_keys: &[&str], exempt_prefixes: &[&str]) -> bool {
    exempt_keys.iter().any(|candidate| key == *candidate)
        || exempt_prefixes.iter().any(|prefix| key.starts_with(prefix))
}

fn join_or_none(values: &[String]) -> String {
    if values.is_empty() {
        "<none>".to_string()
    } else {
        values.join(", ")
    }
}
